### ISSUE TITLE
flutter engine may invoke vkQueueSubmit from the io thread

### ISSUE THREAD

**Issue by eladmaimoni** [flutter/flutter#133933]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

I am receiving a validation error with my custom embedder that uses vulkan. I think this happens when flutter loads image content to a texture though I am not sure.

### Expected results

vkQueueSubmit must only be called on the rendering thread (which is the application main thread in my case)

### Actual results

vkQueueSubmit is sometimes called from the IO thread which is not controlled by the embedder.

Validation Error: [ UNASSIGNED-Threading-MultipleThreads ] Object 0: handle = 0x1b322e39ac0, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x141cb623 | THREADING
 ERROR : vkQueueSubmit(): object of type VkQueue is simultaneously used in thread 55012 and thread 48720

### Code sample

The relevant code is in the flutter engine, here is the callstack:

```
 	ultra.api.dll!ultra::vulkan::DebugMessengerCallback(VkDebugUtilsMessageSeverityFlagBitsEXT messageSeverity, unsigned int __formal, const VkDebugUtilsMessengerCallbackDataEXT * pCallbackData, void * __formal) Line 52	C++
 	VkLayer_khronos_validation.dll!debug_log_msg(const _debug_report_data * debug_data, unsigned int msg_flags, const LogObjectList & objects, const char * message, const char * text_vuid, const char *) Line 209	C++
 	VkLayer_khronos_validation.dll!LogMsg(const _debug_report_data * debug_data, unsigned int msg_flags, const LogObjectList & objects, const std::string & vuid_text, const char * format, char * argptr) Line 445	C++
 	VkLayer_khronos_validation.dll!ValidationObject::LogError(const LogObjectList & objlist, const std::string & vuid_text, const char * format, ...) Line 4045	C++
 	VkLayer_khronos_validation.dll!counter<VkQueue_T *>::StartWrite(VkQueue_T * object, const char * api_name) Line 172	C++
 	[Inline Frame] VkLayer_khronos_validation.dll!ThreadSafety::StartWriteObject(VkQueue_T *) Line 477	C++
 	VkLayer_khronos_validation.dll!ThreadSafety::PreCallRecordQueueSubmit(VkQueue_T * queue, unsigned int submitCount, const VkSubmitInfo * pSubmits, VkFence_T * fence) Line 961	C++
 	VkLayer_khronos_validation.dll!vulkan_layer_chassis::QueueSubmit(VkQueue_T * queue, unsigned int submitCount, const VkSubmitInfo * pSubmits, VkFence_T * fence) Line 1139	C++
 	flutter_engine.dll!submit_to_queue(GrVkGpu * gpu, VkQueue_T * queue, VkFence_T * fence, unsigned int waitCount, VkSemaphore_T * const * waitSemaphores, const unsigned int * waitStages, unsigned int commandBufferCount, VkCommandBuffer_T * const * commandBuffers, unsigned int signalCount, VkSemaphore_T * const * signalSemaphores, skgpu::Protected protectedContext) Line 561	C++
 	flutter_engine.dll!GrVkPrimaryCommandBuffer::submitToQueue(GrVkGpu * gpu, VkQueue_T * queue, skia_private::TArray<GrVkSemaphore::Resource *,1> & signalSemaphores, skia_private::TArray<GrVkSemaphore::Resource *,1> & waitSemaphores) Line 597	C++
 	flutter_engine.dll!GrVkGpu::submitCommandBuffer(GrVkGpu::SyncQueue sync) Line 411	C++
 	flutter_engine.dll!GrGpu::submitToGpu(bool syncCpu) Line 760	C++
 	flutter_engine.dll!GrVkGpu::prepareTextureForCrossContextUsage(GrTexture * texture) Line 2769	C++
 	flutter_engine.dll!SkImages::CrossContextTextureFromPixmap(GrDirectContext * dContext, const SkPixmap & originalPixmap, bool buildMips, bool limitToMaxTextureSize) Line 418	C++
 	[External Code]	
 	[Inline Frame] flutter_engine.dll!std::_Func_class<void>::operator()() Line 874	C++
 	flutter_engine.dll!fml::SyncSwitch::Execute(const fml::SyncSwitch::Handlers & handlers) Line 27	C++
 	[External Code]	
 	[Inline Frame] flutter_engine.dll!std::_Func_class<void>::operator()() Line 874	C++
>	flutter_engine.dll!fml::MessageLoopImpl::FlushTasks(fml::FlushType type) Line 127	C++
 	flutter_engine.dll!fml::MessageLoopWin::Run() Line 39	C++
 	flutter_engine.dll!fml::MessageLoopImpl::DoRun() Line 94	C++
 	[External Code]	
 	flutter_engine.dll!thread_start<unsigned int (__cdecl*)(void *),1>(void * const parameter) Line 97	C++
 	[External Code]	

```

****

### Screenshots or Video

![image](https://github.com/flutter/flutter/assets/17446835/558060b2-7d50-4add-aa26-26ac42e84c08)

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

$ flutter doctor -v
[!] Flutter (Channel [user-branch], 3.12.0-11.0.pre.11, on Microsoft Windows [Version 10.0.19045.3324], locale en-IL)
    ! Flutter version 3.12.0-11.0.pre.11 on channel [user-branch] at C:\libs\flutter
      Currently on an unknown channel. Run `flutter channel` to switch to an official channel.
      If that doesn't fix the issue, reinstall Flutter by following instructions at
      https://flutter.dev/docs/get-started/install.
    ! Upstream repository unknown source is not a standard remote.
      Set environment variable "FLUTTER_GIT_URL" to unknown source to dismiss this error.
    • Framework revision 176f7e6382 (2 months ago), 2023-06-25 00:00:25 -0400
    • Engine revision 100f6fc854
    • Dart version 3.1.0 (build 3.1.0-246.0.dev)
    • DevTools version 2.24.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly
      to perform update checks and upgrades.

[✓] Windows Version (Installed version of Windows is version 10 or higher)

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.5.1)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.5.33424.131
    • Windows 10 SDK version 10.0.22000.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[✓] VS Code, 64-bit edition (version 1.81.1)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.58.0

[✓] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19045.3324]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 114.0.5735.199
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 115.0.1901.203

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 3 categories.

**Comment by gaaclarke** [comment]
I chatted with the Skia team.  They said that the way this works is that the client of Skia is providing the `vkQueue`s and it has no way of knowing that they should be threadsafe calls.  When setting up the `GrVkBackendContext` we should be supplying a callback that will allow us to insert a mutex around submission.

**Comment by gaaclarke** [comment]
Here is where the swap should be happening to a threadsafe variant: https://github.com/gaaclarke/engine/blob/9fff75d9364fb3eafb62240ac1ab04f76ef914cd/vulkan/procs/vulkan_proc_table.cc#L158C5-L158C5

**Comment by gaaclarke** [comment]
https://github.com/flutter/engine/pull/45459 may fix this issue.  The problem is, I don't think there is any way to test it.  @eladmaimoni do you have a way I can test it?

**Comment by eladmaimoni** [comment]
@gaaclarke thanks for all the digging. 

I don't have a minimal reproducable example I can send you. We wrote a custom embedder so we can integrate our vulkan code with flutter. However, If you could provide me with a compatible engine version (git SHA) and SDK version (git SHA) I can try to compile them and test these fixes for you.

**Comment by gaaclarke** [comment]
@eladmaimoni the code is linked in the PR, you can just download it from my branch.  I know it is quite involved to get setup, sorry.  I did talk to the team yesterday and I think there is a way I can test this with the `embedder_unittests` target.  I'm not familiar with that but I plan on giving it a look.

**Comment by eladmaimoni** [comment]
@gaaclarke I haven't yet tried your branch. but, I think I do have a scenario where this is quite reproducible.

I am using flutter_map package to draw a simple map. it seems that when zooming on the map (which probably trigger downloads of new tiles from the web) causes this validation error to trigger quite a lot. 

so to reproduce you can try to run the following dart example with a vulkan backend.

```
class MapView extends StatelessWidget {
  final mapOrigin = ll.LatLng(32.070381, 34.795758);
    return FlutterMap(
      options: MapOptions(
        center: mapOrigin,
        zoom: 15.3,
        maxZoom: 18.4,
      ),
      // nonRotatedChildren: [
      //   RichAttributionWidget(
      //     attributions: [
      //       TextSourceAttribution(
      //         'OpenStreetMap contributors',
      //         onTap: () =>
      //             launchUrl(Uri.parse('https://openstreetmap.org/copyright')),
      //       ),
      //     ],
      //   ),
      // ],
      children: [
        TileLayer(
          urlTemplate: 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
          userAgentPackageName: 'com.example.app',
        ),
          ],
        ),
      ],
    );
  }
}
```

Hope this helps.

**Comment by gaaclarke** [comment]
I'm unable to reproduce the problem directly because we don't have the exact setup and the problem is a race condition.  I was able to exercise the code with tests and based on the stacktrace in the bug report determine the cause.  Based on that information I believe this should be fixed now.  The problem was talking to a vkQueue simultaneously, now there are mutexes guarding against that.

**Comment by eladmaimoni** [comment]
@gaaclarke coming to think about it some more, does this mutex protect against race conditions with the embedder code? after all, the embedder uses the queue to compose the final target and to render platform views via a compositor.

**Comment by gaaclarke** [comment]
> @gaaclarke coming to think about it some more, does this mutex protect against race conditions with the embedder code? after all, the embedder uses the queue to compose the final target and to render platform views via a compositor.

That hadn't occurred to me, thanks.  It might be a problem.  I'd have to see what is actually being called and how.  Do you have code that you could show me how you are using this?  @bdero will have a better idea of how this is supposed to be set up.  At first blush our options are:
1) Pass in a mutex to the embedder
2) Add an accessor to the mutex in the embedder api
3) Move all communication to the raster thread

Options 1 and 2 have bad usability since we can't force correct usage.  Option 3 has runtime costs.

If however users of the embedder api are using our vulkan context with the threadsafe variants, there should be no problem.

**Comment by eladmaimoni** [comment]
@gaaclarke I have my own embedder code which I can theoretically share, though it might be hard to decouple it might take some time to decouple it from the rest of my irrelevant and proprietary code. 

The general idea is this though:

1. embedder initializes vulkan, gives flutter a VkDevice, VkInstance and VkQueue. 
2. embedder intializes a render_task_runner for the render thread.
3. embedder initializes a FlutterCompositor for receiving callbacks to the render thread (via the task runner) to perform composition and presentation.
4. the embedder draws custom platform views and preform presentation when `present_layers_callback` is called.

The bottom line is, there is nothing preventing the embedder from using the VkQueue as it see fit. For example - outside the `present_layers_callback` to prepare some textures. A reasonable assumption is that the VkQueue will be accessed by the render thread only. This is what almost all vulkan code I've seen do.

I think that the simplest solution is for flutter to enqueue a one-time task to the render task runner to perform the command buffer submission. 

Regarding runtime cost - I don't fully understand how this can occur. VkQueueSubmit is an asynchronous call. the actual work is being done by the GPU anyway.

**Comment by gaaclarke** [comment]
Thanks, yes I don't think the PR I landed is going to work.  I'm going to reopen this issue.

> Regarding runtime cost - I don't fully understand how this can occur. VkQueueSubmit is an asynchronous call. the actual work is being done by the GPU anyway.

It introduces latency between when you want to execute `vkQueueSubmit` and when it actually gets executed, more than if you are using a mutex since you are potentially blocked on other work in the raster thread and the swapping of the thread context.

**Comment by eladmaimoni** [comment]
@gaaclarke I get the latency part. Though I believe this is a very reasonable compromise. The render thread should poll / wait on the task runner queue most of the time. the task runner architecture has a mechanism to determine the "urgency" of the task (target time nanos). Just like the render thread may be "late" fo presentation or other urgent tasks, it may be "late" for this one time task.

**Comment by gaaclarke** [comment]
Yea, I'm convinced. I'll give it a try today.  Thanks again.

**Comment by gaaclarke** [comment]
Moving Skia's usage of the vkQueue from the io thread to the raster thread is risky too since it may introduce different race conditions where other Skia internal data is being accessed simultaneously on the raster thread.  This is going to take a bit more time to work with the team to figure out what we want to do here.

Here are some other things I considered:
1) Merge the io thread and raster thread if vulkan embedder is used -- This would lead to deadlocks without a major rewrite.
1) Create task runners that obtain a mutex when executing so that the raster thread and io thread never execute simultaneously -- This would fix the issue but would be very bad for performance, a nuclear option.
1) Expand the embedder api so that users can swap out `vkQueueSubmit` and `vkQueueWaitIdle` for threadsafe variants like I attempted for Skia -- The embedder api has the ability to specify the instance procs via `FlutterVulkanRendererConfig::get_instance_proc_address_callback` but no way to swap out the device procs so users can't work around the issue themselves today.

**Comment by gaaclarke** [comment]
@uysalere I reopened this when that PR was reverted.

**Comment by bdero** [comment]
FYI @jwinarske, in case you've run into issues here.

Just catching up on this, but I'm not sure this problem is embedder API specific. I think we'll want to fix this for our other Vulkan usage in the engine as well by gracefully supporting single queue.

The first question that comes into my mind is: Why is the Android embedder not triggering this validation error when running the Impeller Vulkan backend?

Impeller Vulkan currently uses one graphics queue to submit everything (even though the setup routine has logic to find a dedicated transfer queue, [falling back to the graphics queue](https://github.com/flutter/engine/blob/1369ab35aaa79639f566dbdbbbb5476eda889e31/impeller/renderer/backend/vulkan/context_vk.cc#L272-L275) if one couldn't be found). And the Impeller image decoder performs image uploads on the IO thread not unlike the Skia image decoder.

> The embedder api has the ability to specify the instance procs via FlutterVulkanRendererConfig::get_instance_proc_address_callback but no way to swap out the device procs so users can't work around the issue themselves today.

This can be done by intercepting `GetDeviceProcAddr` in the instance proc callback.

**Comment by eladmaimoni** [comment]
@gaaclarke 

> Moving Skia's usage of the vkQueue from the io thread to the raster thread is risky too since it may introduce different race conditions where other Skia internal data is being accessed simultaneously on the raster thread. This is going to take a bit more time to work with the team to figure out what we want to do here.

What about blocking the io thread until the raster thread completes the task? I don't know if the io thread is allowed to perform blocking io, but if it does, it might be a solution. 

this is of course assuming you could obtain a fence / semaphore from skia that can you can wait on.

**Comment by gaaclarke** [comment]
> Just catching up on this, but I'm not sure this problem is embedder API specific. I think we'll want to fix this for our other Vulkan usage in the engine as well by gracefully supporting the single queue case.
> 
> The first question that comes into my mind is: Why is the Android embedder not triggering this validation error when running the Impeller Vulkan backend?

In impeller we have a [mutex around the usage of the queues]( https://github.com/flutter/engine/blob/d4698c65aa8d9e42d4aa70391c804270506a5f82/impeller/renderer/backend/vulkan/queue_vk.cc#L22).  Impeller today has full ownership of the queue so it doesn't have the same problems that the Skia/Embedder setup will have.  It will be a problem when we move embedder to impeller.

> This can be done by intercepting `GetDeviceProcAddr` in the instance proc callback.

Oh excellent, so there is a workaround today.  I wish we should have made the proc callback like skia's which combines these into one.

## Workaround

Insert your own mutex around queue usage into the flutter engine.

1) specify `FlutterVulkanRendererConfig::get_instance_proc_address_callback` to return your own function for `vkGetDeviceProcAddr`
1) In your `vkGetDeviceProcAddr` override `vkQueueSubmit` and `vkQueueWaitIdle` to return threadsafe variants that share a mutex (see https://github.com/flutter/engine/pull/45459 for an example on how to do that)
1) In the embedder host code, use the same mutex around the vkQueue usage that was used in the `vkQueueSubmit` and `vkQueueWaitIdle` threadsafe variants.

**Comment by bdero** [comment]
> In impeller we have a [mutex around the usage of the queues](https://github.com/flutter/engine/blob/d4698c65aa8d9e42d4aa70391c804270506a5f82/impeller/renderer/backend/vulkan/queue_vk.cc#L22).

Thanks, that's what I was missing.

It sounds like you might be on the same page, but my suggestion for solving this would be something like:
1. Add an engine setting "upload_on_raster_thread" or similar that activates when using the Vulkan embedder API.
2. When this flag is enabled, do uploads in a raster task instead of an IO task (may require cleaning out some cobwebs). I'm also happy to help if you're not up for it. Need a GPU sync switch accessible on the raster thread.
3. Later on, we can add an optional `transfer_queue` field to `embedder.h` that explicitly guarantees use on the I/O thread only. This is an easy improvement and is how most embedders should use the API for perf reasons anyhow.

We could also just roll with the `transfer_queue` solution right away, but not also supporting the single queue approach would technically be an API break (and there may be devices out there that don't support multiple queues).

**Comment by bdero** [comment]
I filed follow-up issues about this. @eladmaimoni could you try the lock-based workaround mentioned in https://github.com/flutter/flutter/issues/133933#issuecomment-1716031565 and see if that works for you?

We definitely need to fix this internal behavior eventually, but it's not high on our list of priorities at the moment. Once we resolve https://github.com/flutter/flutter/issues/134573, you'll be able to remove the workaround from your custom embedder.

**Comment by eladmaimoni** [comment]
@gaaclarke @bdero 

I implemented the workaround and can confirm it is working. 

Though I don't use a mutex. I simply pass the call to vkQueueSubmit to the main thread and wait on a future:

```
    thread_local bool is_main_thread = false;
    VkDevice g_device = nullptr;
 
    concurrencpp::result<VkResult> EnqueueSubmitOnMainThread(VkQueue queue, uint32_t submitCount, const VkSubmitInfo* pSubmits, VkFence fence)
    {
        co_await PreRenderRecordResume(); // switch to main thread

        co_return vkQueueSubmit(queue, submitCount, pSubmits, fence);
    }

    VkResult FlutterQueueSubmit(VkQueue queue, uint32_t submitCount, const VkSubmitInfo* pSubmits, VkFence fence)
    {
        if (!is_main_thread) // thread_local variable
        {
            // HACK around https://github.com/flutter/flutter/issues/133933
            auto future = EnqueueSubmitOnMainThread(queue, submitCount, pSubmits, fence);
            return future.get();
        }

        return vkQueueSubmit(queue, submitCount, pSubmits, fence);
    }


    void* FlutterGetDeviceProcAddr(VkDevice device, const char* pName)
    {
        if (0 == std::strcmp(pName, "vkQueueSubmit"))
        {
            return FlutterQueueSubmit;
        }
        return vkGetDeviceProcAddr(device, pName);
    }

    void* FlutterGetInstanceProcAddr(VkInstance instance, const char* pName)
    {
        if (0 == std::strcmp(pName, "vkGetDeviceProcAddr"))
        {
            return FlutterGetDeviceProcAddr;
        }
        return vkGetInstanceProcAddr(instance, pName);
    }
```

Thanks for all your help! much appreciated!
